pkg/secrets-store/utils.go

```diff
@@ -0,0 +1,727 @@
+			notMnt = false
+			return !notMnt, err
+		}
+	}
+
+	return false, nil
+}
+
+// syncK8sObjects creates or updates K8s secrets based on secretProviderClass spec and data from mounted files in targetPath
+// it should also add pod info to the secretProviderClass object's byPod status field
+func syncK8sObjects(ctx context.Context, targetPath string, podUID string, namespace string, secretProviderClass string, secretObjects []interface{}) error {
+	successfulUpdates := 0
+	files, err := getMountedFiles(targetPath)
+	if err != nil {
+		return err
+	}
+	for _, s := range secretObjects {
+		secretObject, ok := s.(map[string]interface{})
+		if !ok {
+			continue
+		}
+		secretName, exists, err := unstructured.NestedString(secretObject, "secretName")
+		if err != nil {
+			return err
+		}
+		if !exists {
+			continue
+		}
+		sType, exists, err := unstructured.NestedString(secretObject, "type")
+		if err != nil {
+			return err
+		}
+		if !exists {
+			log.Infof("type does not exist in obj")
+			continue
+		}
+		secretType := getSecretType(sType)
+		secretObjectDataList, exists, err := unstructured.NestedSlice(secretObject, "data")
+		if err != nil {
+			return err
+		}
+		if !exists {
+			continue
+		}
+		datamap := make(map[string][]byte)
+		for _, d := range secretObjectDataList {
+			secretObjectData, ok := d.(map[string]interface{})
+			if !ok {
+				continue
+			}
+			objectName, exists, err := unstructured.NestedString(secretObjectData, "objectName")
+			if err != nil {
+				return err
+			}
+			if !exists {
+				continue
+			}
+			key, exists, err := unstructured.NestedString(secretObjectData, "key")
+			if err != nil {
+				return err
+			}
+			if !exists {
+				continue
+			}
+			found := false
+			for _, file := range files {
+				filename := filepath.Base(file)
+				if filename == objectName {
+					found = true
+					log.Infof("file matching objectName %s found, processing key %s", objectName, key)
+					data, err := ioutil.ReadFile(file)
+					if err != nil {
+						log.Errorf("failed to read file %s, err: %v", file, err)
+						return status.Error(codes.Internal, err.Error())
+					}
+					if secretType == corev1.SecretTypeTLS {
+						data, err = getCertPart(data, key)
+						if err != nil {
+							log.Errorf("failed to get cert data from file %s, err: %v", file, err)
+							return status.Error(codes.Internal, err.Error())
+						}
+					}
+					datamap[key] = data
+					break
+				}
+			}
+			if !found {
+				log.Errorf("file matching objectName %s not found", objectName)
+			}
+		}
+		createFn := func() (bool, error) {
+			if err := createOrUpdateK8sSecret(ctx, secretName, namespace, datamap, secretType); err != nil {
+				return false, nil
+			}
+			successfulUpdates++
+			return true, nil
+		}
+		if err := wait.ExponentialBackoff(wait.Backoff{
+			Steps:    5,
+			Duration: 1 * time.Millisecond,
+			Factor:   1.0,
+			Jitter:   0.1,
+		}, createFn); err != nil {
+			log.Error(err, "max retries for creating secret reached")
```

Comment:
we should wrap this error with secret name. The error from `wait.ExponentialBackoff` will be `timed out waiting for the condition`